File: courses/codes/food_finder.py
def dfs(G, food_locs, C, dest_loc):
    V = dict(zip(list(G.keys()), [False] * len(G)))
    S = []

    # keep track of how many food have been collected
    # and also if the destination is reachable
    num_of_food_obtained = 0
    dest_reached = False

    node = (0,0)            # always start at (0,0)
    S.append(node)
    while len(S) > 0:
        node = S.pop(-1)
        if not V[node]:
            V[node] = True

            # keep track of how much food collected and if the destination has been reached
            if node in food_locs:
                num_of_food_obtained += 1
            if node == dest_loc:
                dest_reached = True

            for nbr in G[node]:
                if not V[nbr]:
                    S.append(nbr)

    return num_of_food_obtained >= C and dest_reached

# helper function only ("cross product")
# example input List: {(0,0), (1,1), (2,2)}
# example output List: {((0,0), (1,1)), ((0,0), (2,2)), ((1,1), (2,2))}
def get_pairs(L):
    pairs = []
    for l1 in L:
        for l2 in L:
            pairs.append((l1, l2))
    return pairs

C = int(input())    # amount of food required
m = int(input())    # rows
n = int(input())    # cols

food_loc = []
G = dict()
for i in range(m):
    line = input()
    for j in range(n):
        if line[j] != 'X':
            loc = (i, j)
            if loc not in G:
                G[loc] = []
            if line[j] == 'F':
                food_loc.append(loc)

nodes = list(G.keys())
for u, v in get_pairs(nodes):
    if u != v:
        if v[1] == u[1] and abs(v[0] - u[0]) == 1 or v[0] == u[0] and abs(v[1] - u[1]) == 1:
            G[u].append(v)
            G[v].append(u)

# remove duplicates in the adj list:
for node in G.keys():
    val = list(set(G[node]))
    G[node] = val

output = 'Y' if dfs(G, food_loc, C, (m-1, n-1)) else 'N'
print(output)

File: courses/codes/FoodFinder.java
import java.util.*;
import java.lang.Math;

public class FoodFinder {
	
	// create a helper class called Pair to handle ordered pairs
	public static class Pair<X, Y> {
        public final X x;
        public final Y y;
        
        public Pair(X x, Y y) {
            this.x = x;
            this.y = y;
        }
        
        public void print() {
            System.out.print( "(" + this.x + ", " + this.y + ")" );
        }
        
        public void print(String s) {
            System.out.println( "(" + this.x + ", " + this.y + ")" );
        }
        
        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (other == null || getClass() != other.getClass()) return false;
            Pair<X, Y> pair = (Pair<X, Y>) other;
            return x.equals(pair.x) && y.equals(pair.y);
        }

        @Override
        public int hashCode() {
            int result = x != null ? x.hashCode() : 0;
            result = 31 * result + (y != null ? y.hashCode() : 0);
            return result;
        }
    }
	 
	public static boolean dfs(Map<Pair<Integer, Integer>, List<Pair<Integer, Integer>>> G, List<Pair<Integer, Integer>> foodLocs, Integer C, Pair<Integer, Integer> destLoc) {
		Stack<Pair<Integer, Integer>> S = new Stack<>();
		Map<Pair<Integer, Integer>, Boolean> V = new HashMap<>();
		for (Pair<Integer, Integer> node : G.keySet()) {
			V.put(node, false);
		}
		
		int numOfFoodObtained = 0;
		boolean destReached = false;
		S.push(new Pair<>(0,0));
		
		while (!S.isEmpty()) {
			Pair<Integer, Integer> node = S.pop();
			if (!V.get(node)) {
				V.put(node, true);
				
				if (foodLocs.contains(node)) {
					numOfFoodObtained++;
				}
				if (node.equals(destLoc)) {
					destReached = true;
				}
				
				for (Pair<Integer, Integer> nbr : G.get(node)) {
					if (!V.get(nbr)) {
						S.push(nbr);
					}
				}
			}
			
		}
		
		return numOfFoodObtained >= C && destReached;
		
	}
	
    // helper function only ("cross product")
    // example input List: {(0,0), (1,1), (2,2)}
    // example output List: {((0,0), (1,1)), ((0,0), (2,2)), ((1,1), (2,2))}
	public static List<Pair<Pair<Integer, Integer>, Pair<Integer, Integer>>> getPairList(List<Pair<Integer, Integer>> nodes) {
		List<Pair<Pair<Integer, Integer>, Pair<Integer, Integer>>> pairs = new ArrayList<>();
		for (int i = 0; i < nodes.size(); i++) {
			for (int j = i; j < nodes.size(); j++) {
				Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> pair = new Pair<>(nodes.get(i), nodes.get(j));
				pairs.add(pair);
			}
		}
		return pairs;
	}
	
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int C = Integer.parseInt(sc.nextLine());	// amount of food required
		int m = Integer.parseInt(sc.nextLine());	// rows
		int n = Integer.parseInt(sc.nextLine());	// cols
		
		// locations with food
		List<Pair<Integer, Integer>> foodLoc = new ArrayList<>();
		
		// Create adj list
		Map<Pair<Integer, Integer>, List<Pair<Integer, Integer>>> G = new HashMap<>();
		
		for (int i = 0; i < m; i++) {
			String line = sc.nextLine();
			for (int j = 0; j < n; j++) {
				if (line.charAt(j) != 'X') {
					Pair<Integer, Integer> loc = new Pair<>(i, j);
					if (!G.containsKey(loc)) {
						G.put(loc, new ArrayList<>());
					}
					if (line.charAt(j) == 'F') {
						foodLoc.add(loc);
					}
				}
			}
		}
		
		List<Pair<Integer, Integer>> my_nodes = new ArrayList<>();
		my_nodes.addAll(G.keySet());
		
		for (Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> combo : getPairList(my_nodes)) {
			Pair<Integer, Integer> u = combo.x;
			Pair<Integer, Integer> v = combo.y;
			if (!u.equals(v)) {
				if ((v.x == u.x && Math.abs(v.y - u.y) == 1) || (v.y == u.y && Math.abs(v.x - u.x) == 1)) {
					List<Pair<Integer, Integer>> nbrs = G.get(u);
					nbrs.add(v);
					G.put(u, nbrs);
					
					nbrs = G.get(v);
					nbrs.add(u);
					G.put(v, nbrs);
				}
			}
		}
		
		Pair<Integer, Integer> dest = new Pair<>(m-1, n-1);
		char output = dfs(G, foodLoc, C, dest) ? 'Y' : 'N';
		System.out.println(output);
		
		sc.close();
	}


}